ocaml/xapi/xapi_database_backup.ml

```diff
@@ -0,0 +1,357 @@
+
+let get_delta __context token =
+  Mutex.execute Db_lock.dbcache_mutex (fun () ->
+      let db = Db_ref.get_database (Context.database_of __context) in
+      let tables = check_for_updates token db in
+      let deleted = get_deleted token db in
+      let t = Manifest.generation (Database.manifest db) in
+      {fresh_token=t;
+       last_event_token=Int64.pred t;
+       tables=tables;
+       deletes=deleted;
+       counts=object_count db})
+
+let handler (req: Request.t) s _ =
+  try
+    let token = Generation.of_string (snd (List.find (fun (x,y) -> x = "token") req.query)) in
+    Server_helpers.exec_with_new_task "Database backup" ~task_in_database:false (fun __context ->
+        token
+        |> get_delta __context
+        |> rpc_of_delta
+        |> Jsonrpc.to_string
+        |> Http_svr.response_str req s
+      )
+  with Not_found -> debug "No valid token found in request"
+
+(* Slave Functions *)
+
+let apply time tblname objref fldname newval =
+  (fun _ -> newval)
+  |> Row.update time fldname newval
+  |> Table.update time objref Row.empty
+  |> TableSet.update time tblname Table.empty
+  |> Database.update
+
+let make_from_barrier stat tblname objref =
+  Row.empty
+  |> Table.touch stat.modified objref
+  |> TableSet.update stat.modified tblname Table.empty
+  |> Database.update
+
+let make_empty_table stat tblname =
+  (fun t -> t)
+  |> TableSet.update stat.created tblname Table.empty
+  |> Database.update
+
+let make_change stat tblname objref fldname newval db =
+  if TableSet.mem tblname (Database.tableset db) then begin
+    if (Table.mem objref (TableSet.find tblname (Database.tableset db))) then begin
+      if (Row.mem fldname (Table.find objref (TableSet.find tblname (Database.tableset db)))) then
+        apply stat.modified tblname objref fldname newval db
+      else
+        begin
+          if stat.created = stat.modified then
+            apply stat.created tblname objref fldname newval db
+          else
+            apply stat.modified tblname objref fldname newval (apply stat.created tblname objref fldname (Schema.Value.String "XXX") db)
+        end
+    end
+    else
+      begin
+        if stat.created = stat.modified then
+          apply stat.created tblname objref fldname newval db
+        else
+          apply stat.modified tblname objref fldname newval (apply stat.created tblname objref fldname (Schema.Value.String (tblname^objref)) db)
+      end
+  end
+  else
+    apply stat.modified tblname objref fldname newval (make_empty_table {stat with created=(-1L);} tblname db)
+
+
+let compare_by_tblname (c1: (string * int)) (c2: (string * int)) =
+  String.compare (fst c1) (fst c2)
+
+let count_check (c1_list: (string * int) list) (c2_list: (string * int) list) =
+  let c1_list = List.sort compare_by_tblname c1_list in
+  let c2_list = List.sort compare_by_tblname c2_list in
+  try
+    List.for_all2 (fun (a,b) (c,d) -> a=c && b=d) c1_list c2_list
+  with invalid_arg -> false
```

Comment:
`invalid_arg` is an identifier here that binds an exception and not an exception itself - because it is lowercase. Is this intended?